File: 2014-10-10-FP/PrefixCompression/pc.accepted.hs
slv "" ys = ("", "", ys)
slv xs "" = ("", xs, "")
slv xs'@(x : xs) ys'@(y : ys) =
    if x == y
        then
            let (a', b', c') = slv xs ys
                in (x : a', b', c')
        else ("", xs', ys')

main = do
    astr <- getLine
    bstr <- getLine
    let (a, b, c) = slv astr bstr
    putStrLn $ show (length a) ++ " " ++ a
    putStrLn $ show (length b) ++ " " ++ b
    putStrLn $ show (length c) ++ " " ++ c

File: 2014-10-10-FP/PrefixCompression/pc.accepted.fs
open System

let rec slv =
    function
    | (pfx, [], y) -> (pfx, [], y)
    | (pfx, x, []) -> (pfx, x, [])
    | (pfx, ((x :: xs') as xs), ((y :: ys') as ys)) ->
        if x = y
            then slv (x :: pfx, xs', ys')
            else (pfx, xs, ys)

let tostr x = System.String (List.toArray x)

let main _ =
    let astr = List.ofSeq (Console.ReadLine ())
    let bstr = List.ofSeq (Console.ReadLine ())
    let (a, b, c) = slv ([], astr, bstr)
    Console.WriteLine ("{0} {1}", a.Length, tostr (List.rev a))
    Console.WriteLine ("{0} {1}", b.Length, tostr b)
    Console.WriteLine ("{0} {1}", c.Length, tostr c)

main ()

